std/experimental/allocator/building_blocks/aligned_block_list.d

```diff
@@ -0,0 +1,629 @@
+    }
+
+    static if (hasMember!(ParentAllocator, "alignedAllocate"))
+    void[] allocate(size_t n)
+    {
+        static if (isShared)
+        import core.atomic : atomicOp, atomicLoad;
+
+        if (n == 0 || n > alignment)
+            return null;
+
+        static if (isShared)
+        {
+            lock.lock();
+            scope(exit) lock.unlock();
+        }
+
+        auto tmp = cast(AlignedBlockNode*) root;
+
+        // Iterate through list and find first node which has memory available
+        while (tmp)
+        {
+            auto next = tmp.next;
+            static if (isShared)
+            {
+                // Make sure nobody deletes this node while using it
+                tmp.keepAlive++;
+                if (next) next.keepAlive++;
+                lock.unlock();
+            }
+
+            auto result = tmp.bAlloc.allocate(n);
+            if (result.length == n)
+            {
+                static if (isShared)
+                {
+                    atomicOp!"+="(tmp.bytesUsed, n);
+                    lock.lock();
+                }
+                else
+                {
+                    tmp.bytesUsed += n;
+                }
+
+                moveToFront(tmp);
+
+                static if (isShared)
+                {
+                    tmp.keepAlive--;
+                    if (next) next.keepAlive--;
+                }
+
+                return result;
+            }
+
+            // This node can now be removed if necessary
+            static if (isShared)
+            {
+                lock.lock();
+                tmp.keepAlive--;
+                if (next) next.keepAlive--;
+            }
+
+            if (!next)
+                break;
+
+            tmp = next;
+            next = tmp.next;
+
+            // If there are too many nodes, free memory by removing empty nodes
+            static if (isShared)
+            {
+                if (atomicLoad(numNodes) > maxNodes &&
+                    atomicLoad(tmp.bytesUsed) == 0 &&
+                    tmp.keepAlive == 0)
+                {
+                    removeNode(tmp);
+                }
+            }
+            else
+            {
+                if (numNodes > maxNodes && tmp.bytesUsed == 0)
+                {
+                    removeNode(tmp);
+                }
+            }
+
+            tmp = next;
+        }
+
+        // Cannot create new AlignedBlockNode. Most likely the ParentAllocator ran out of resources
+        if (!insertNewNode())
+            return null;
+
+        tmp = cast(typeof(tmp)) root;
+        void[] result = tmp.bAlloc.allocate(n);
+
+        static if (isShared)
+        {
+            atomicOp!"+="(root.bytesUsed, result.length);
+        }
+        else
+        {
+            root.bytesUsed += result.length;
+        }
+
+        return result;
+    }
+
+    size_t goodAllocSize(const size_t n)
+    {
+        Allocator a = null;
+        return a.goodAllocSize(n);
```

Comment:
Wouldn't this segfault?

Reply:
If goodAllocSize does not use internal allocator state (and it shouldn't) it won't segfault.
Currently, this would work with all allocators in experimental